src/mobu/services/business/nublado.py

```diff
@@ -310,89 +274,55 @@ async def _spawn_lab(self, sw: Stopwatch) -> bool:  # noqa: C901
             raise
-        except Exception as e:
-            log = "\n".join([str(m) for m in log_messages])
-            user = self.user.username
-            raise JupyterSpawnError.from_exception(
-                log,
-                e,
-                user,
-                event=sw.event,
-                annotations=sw.annotations,
-                started_at=sw.start_time,
-            ) from e
 
         # We only fall through if the spawn failed, timed out, or if we're
         # stopping the business.
         if self.stopping:
             return False
         log = "\n".join([str(m) for m in log_messages])
-        if sw.elapsed > timeout:
-            elapsed_seconds = round(sw.elapsed.total_seconds())
-            msg = f"Lab did not spawn after {elapsed_seconds}s"
-            raise JupyterTimeoutError(
-                msg,
-                self.user.username,
-                log,
-                event=sw.event,
-                started_at=sw.start_time,
-            )
-        raise JupyterSpawnError(
-            log,
-            self.user.username,
-            event=sw.event,
-            started_at=sw.start_time,
+        sentry_sdk.get_current_scope().add_attachment(
+            filename="spawn_log.txt",
+            bytes=remove_ansi_escapes(log).encode(),
         )
+        spawn_duration = duration(span)
+        if spawn_duration > timeout:
+            elapsed_seconds = round(spawn_duration.total_seconds())
+            msg = f"Lab did not spawn after {elapsed_seconds}s"
+            raise JupyterSpawnTimeoutError(msg)
+        raise JupyterSpawnError
 
     async def lab_login(self) -> None:
         self.logger.info("Logging in to lab")
-        with self.timings.start("lab_login", self.annotations()) as sw:
-            try:
-                await self._client.auth_to_lab()
-            except ne.JupyterProtocolError as exc:
-                raise JupyterProtocolError.from_client_exception(
-                    exc,
-                    event=sw.event,
-                    annotations=sw.annotations,
-                    started_at=sw.start_time,
-                ) from exc
+        with start_span(op="lab_login"):
+            await self._client.auth_to_lab()
 
     @asynccontextmanager
     async def open_session(
         self, notebook: str | None = None
     ) -> AsyncIterator[JupyterLabSession]:
         self.logger.info("Creating lab session")
         opts = {"max_websocket_size": self.options.max_websocket_message_size}
-        stopwatch = self.timings.start("create_session", self.annotations())
+        create_session_cm = start_span(op="create_session")
+        create_session_cm.__enter__()
         async with self._client.open_lab_session(notebook, **opts) as session:
-            stopwatch.stop()
-            with self.timings.start("execute_setup", self.annotations()):
+            create_session_cm.__exit__(None, None, None)
+            with start_span(op="execute_setup"):
```

Comment:
This is pretty ugly. I think it would be better to wrap the Nublado client instead (or, maybe better, add Sentry support directly to the Nublado client?) so that the span management happens inside the async context manager returned by `open_lab_session`.

Reply:
Yeah, I agree that this should probably be in the nublado client itself, and I can add a ticket to do that. This is just a "don't make it worse" direct conversion for now :(

Reply:
Yup, this looks fine for now.